packages/engine/tests/engine/input/systems/ClientInputSystem.test.tsx

```diff
@@ -0,0 +1,325 @@
+  })
+
+  after(() => {
+    globalThis.ReferenceSpace.origin = originCopy
+  })
+
+  beforeEach(() => {
+    createEngine()
+  })
+
+  afterEach(() => {
+    return destroyEngine()
+  })
+
+  it('test client input system reactor', async () => {
+    const Reactor = SystemDefinitions.get(ClientInputSystem)!.reactor!
+    const { rerender, unmount } = render(<Reactor />)
+    unmount()
+  })
+
+  it('test client input system execute', async () => {
+    const mockXRInputSource = new MockXRInputSource({
+      handedness: 'left',
+      targetRayMode: 'screen',
+      targetRaySpace: new MockXRSpace() as XRSpace,
+      gripSpace: undefined,
+      gamepad: undefined,
+      profiles: ['test'],
+      hand: undefined
+    }) as XRInputSource
+
+    const mockXRFrame = new MockXRFrame()
+    mockXRFrame.pose.transform.position.x = 0.2134
+    mockXRFrame.pose.transform.position.y = 0.3456
+    mockXRFrame.pose.transform.position.z = 0.2345
+    mockXRFrame.pose.transform.orientation.x = 0.4567
+    mockXRFrame.pose.transform.orientation.y = 0.8455
+    mockXRFrame.pose.transform.orientation.z = 0.2454
+    mockXRFrame.pose.transform.orientation.w = 0.2743
+
+    const entity = Engine.instance.originEntity
+    registerState(XRState)
+    getMutableState(XRState).xrFrame.set(mockXRFrame as unknown as XRFrame)
+    setComponent(entity, InputSourceComponent, { source: mockXRInputSource })
+    setComponent(entity, XRSpaceComponent, new MockXRSpace() as XRSpace)
+    setComponent(entity, TransformComponent)
+    const execute = SystemDefinitions.get(ClientInputSystem)!.execute!
+    execute()
+
+    const transformComponent = getComponent(entity, TransformComponent)
+    assert(transformComponent.position.x === mockXRFrame.pose.transform.position.x)
+    assert(transformComponent.position.y === mockXRFrame.pose.transform.position.y)
+    assert(transformComponent.position.z === mockXRFrame.pose.transform.position.z)
+    assert(transformComponent.rotation.x === mockXRFrame.pose.transform.orientation.x)
+    assert(transformComponent.rotation.y === mockXRFrame.pose.transform.orientation.y)
+    assert(transformComponent.rotation.z === mockXRFrame.pose.transform.orientation.z)
+    assert(transformComponent.rotation.w === mockXRFrame.pose.transform.orientation.w)
+  })
+
+  it('test client input system XRUI heuristic', async () => {
+    const mockXRInputSource = new MockXRInputSource({
+      handedness: 'left',
+      targetRayMode: 'screen',
+      targetRaySpace: new MockXRSpace() as XRSpace,
+      gripSpace: undefined,
+      gamepad: undefined,
+      profiles: ['test'],
+      hand: undefined
+    }) as XRInputSource
+
+    const entity = Engine.instance.originEntity
+    registerState(XRUIState)
+    getMutableState(XRUIState).interactionRays.set([
+      new Ray(new Vector3(0.23, 0.65, 0.98), new Vector3(0.21, 0.43, 0.82))
+    ])
+    setComponent(entity, XRUIComponent)
+    XRUIComponent.valueMap[entity] = {
```

Comment:
we should avoid setting valueMap directly, and just pass this object as the property in setComponent